File: js_port/fs_js/src/funcs/list/ReduceListFunction.js
const { BaseFunction, CallType } = require('../../core/functionBase');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');
const { ArrayFsList } = require('../../model/FsList');

class ReduceListFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'Reduce';
    this.callType = CallType.Dual;
  }

  get maxParameters() {
    return 3;
  }

  evaluate(provider, parameters) {
    if (parameters.count !== 2 && parameters.count !== 3) {
      return helpers.makeError(
        helpers.FsError.ERROR_PARAMETER_COUNT_MISMATCH,
        `${this.symbol}: expected 2 or 3 got ${parameters.count}`
      );
    }

    const list = helpers.ensureList(parameters.getParameter(provider, 0));
    if (!list) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The first parameter should be ${this.parName(0)}`
      );
    }

    const fn = helpers.ensureFunction(parameters.getParameter(provider, 1));
    if (!fn) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The second parameter should be ${this.parName(1)}`
      );
    }

    let total;
    let indexStart = 0;

    if (parameters.count === 3) {
      total = parameters.getParameter(provider, 2);
    } else {
      if (list.length === 0) {
        return helpers.typedNull();
      }
      total = list.get(0);
      indexStart = 1;
    }

    for (let i = indexStart; i < list.length; i += 1) {
      const args = new helpers.ArrayParameterList([
        total,
        list.get(i),
        helpers.makeValue(FSDataType.Integer, i)
      ]);
      total = fn.evaluate(provider, args);
    }

    return helpers.ensureTyped(total);
  }

  parName(index) {
    return index === 0 ? 'List' : index === 1 ? 'Transform Function' : 'Seed';
  }
}

module.exports = {
  ReduceListFunction
};

File: Walya/Functions/List/ReduceListFunction.cs
using Walya.Core;
using Walya.Model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Walya.Functions.List
{
    public class ReduceListFunction : IFsFunction
    {
        public int MaxParsCount => 3;

        public CallType CallType => CallType.Dual;

        public string Symbol => "Reduce";

        public int Precedence => 0;
        class DoListFuncPar : IParameterList
        {
            public object S;
            public object X;
            public object I;

            public override int Count => 3;

            public override object GetParameter(IFsDataProvider provider, int index)
            {
                return index switch
                {
                    0 => X,
                    1 => S,
                    2 => I,
                    _ => null,
                };
            }
        }

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            var par0 = pars.GetParameter(parent,0);
            var par1 = pars.GetParameter(parent,1);
            var par2 = pars.GetParameter(parent,2);
            return EvaluateInternal(parent, par0, par1, par2);
        }

        private object EvaluateInternal(IFsDataProvider parent, object par0, object par1, object par2)
        {
            if (par0 == null)
                return null;

            if (!(par0 is FsList))
                throw new Error.TypeMismatchError($"{this.Symbol} function: The first parameter should be {this.ParName(0)}");

            var func = par1 as IFsFunction;

            if (func == null)
                throw new Error.TypeMismatchError($"{this.Symbol} function: The second parameter didn't evaluate to a function");


            var total = par2;

            
            var lst = (FsList)par0;

            for (int i = 0; i < lst.Length; i++)
            {
                
                total = func.Evaluate(parent, new DoListFuncPar { S = total, X = lst[i], I = i });
            }

            return Walya.NormalizeDataType(total);
        }

        public string ParName(int index)
        {
            switch(index)
            {
                case 0:
                    return "List";
                case 1:
                    return "Transform Function";
                default:
                    return "";
            }
        }

    }
}
